File: npc/topics.go
// Package npc：NPC 間對話主題劇本（載入與查詢），供觸發時帶入 AI 或 fallback。
package npc

import (
	"encoding/json"
	"log"
	"math/rand"
	"os"
	"strings"
	"sync"
)

// NpcNpcTopic 單一主題：id、給 AI 的 hint、候選句（可作 fallback 或種子）。
type NpcNpcTopic struct {
	ID           string   `json:"id"`
	Hint         string   `json:"hint"`
	Lines        []string `json:"lines"`
	RequiresWork bool     `json:"requires_work,omitempty"`
	NightOnly    bool     `json:"night_only,omitempty"`
	FollowUp     bool     `json:"follow_up,omitempty"`
}

type npcNpcTopicsFile struct {
	Topics []NpcNpcTopic `json:"topics"`
}

var (
	npcTopicsMu   sync.RWMutex
	npcTopicsList []NpcNpcTopic
)

// LoadNpcNpcTopics 從 path 載入主題劇本，僅執行一次或重載時呼叫。
func LoadNpcNpcTopics(path string) {
	data, err := os.ReadFile(path)
	if err != nil {
		if !os.IsNotExist(err) {
			log.Printf("[npc_topics] load %s: %v", path, err)
		}
		return
	}
	var f npcNpcTopicsFile
	if err := json.Unmarshal(data, &f); err != nil {
		log.Printf("[npc_topics] parse %s: %v", path, err)
		return
	}
	npcTopicsMu.Lock()
	npcTopicsList = f.Topics
	if npcTopicsList == nil {
		npcTopicsList = []NpcNpcTopic{}
	}
	npcTopicsMu.Unlock()
}

// GetNpcNpcTopicByID 依 id 回傳主題；無則回傳 nil。
func GetNpcNpcTopicByID(id string) *NpcNpcTopic {
	npcTopicsMu.RLock()
	defer npcTopicsMu.RUnlock()
	for i := range npcTopicsList {
		if npcTopicsList[i].ID == id {
			return &npcTopicsList[i]
		}
	}
	return nil
}

// normalizeNpcNpcTopicHint 去掉玩家在場時加在 hint 後的括注，便於與劇本 Hint 比對。
func normalizeNpcNpcTopicHint(h string) string {
	h = strings.TrimSpace(h)
	if i := strings.Index(h, "（路人在旁"); i >= 0 {
		h = strings.TrimSpace(h[:i])
	}
	return h
}

// FindNpcNpcTopicIDByHint 依目前 topicHint 反查主題 id（與 thread 存的 TopicType、加註後的 hint 對齊）。
// 先精確比對劇本 Hint，再取「hint 以前綴開頭」中最長的一筆，避免 thread 只存縮寫時無法辨識。
func FindNpcNpcTopicIDByHint(topicHint string) string {
	h := normalizeNpcNpcTopicHint(topicHint)
	if h == "" {
		return ""
	}
	npcTopicsMu.RLock()
	defer npcTopicsMu.RUnlock()
	for i := range npcTopicsList {
		th := strings.TrimSpace(npcTopicsList[i].Hint)
		if th != "" && h == th {
			return npcTopicsList[i].ID
		}
	}
	var bestID string
	var bestLen int
	for i := range npcTopicsList {
		th := strings.TrimSpace(npcTopicsList[i].Hint)
		if th == "" {
			continue
		}
		if strings.HasPrefix(h, th) && len(th) > bestLen {
			bestLen = len(th)
			bestID = npcTopicsList[i].ID
		}
	}
	return bestID
}

// PickRandomNpcNpcTopic 隨機回傳一主題；無主題時回傳 nil。
func PickRandomNpcNpcTopic() *NpcNpcTopic {
	return PickRandomNpcNpcTopicExclude("")
}

// PickRandomNpcNpcTopicExclude 隨機回傳一主題，但排除 excludeID（空字串表示不排除）。用於非職業場所不提示「交班」。
func PickRandomNpcNpcTopicExclude(excludeID string) *NpcNpcTopic {
	npcTopicsMu.RLock()
	list := npcTopicsList
	npcTopicsMu.RUnlock()
	var candidates []int
	for i := range list {
		if excludeID != "" && list[i].ID == excludeID {
			continue
		}
		candidates = append(candidates, i)
	}
	if len(candidates) == 0 {
		return nil
	}
	return &list[candidates[rand.Intn(len(candidates))]]
}

// NpcTopicMask 為主題篩選條件（P2）。
type NpcTopicMask struct {
	IsWorkVenue  bool
	IsNightTime  bool
	HasRoomEvent bool
	// RoomTags 為當前房間 tags（如 JSON 內 "outdoor"），供「城外聞」「天候」等主題加權。
	RoomTags []string
}

// roomHasTag 檢查房間是否含某 tag（與 data/rooms 內英文標籤一致，如 outdoor）。
func roomHasTag(roomTags []string, want string) bool {
	for _, v := range roomTags {
		if v == want {
			return true
		}
	}
	return false
}

// topicMaskBaseWeight 依情境與房間 tag 計算基礎權重（不含雙人關係加成）。
func topicMaskBaseWeight(mask NpcTopicMask, t *NpcNpcTopic) int {
	w := 1
	if mask.IsWorkVenue && t.RequiresWork {
		w += 4
	}
	if mask.HasRoomEvent && t.FollowUp {
		w += 3
	}
	if mask.IsNightTime && t.NightOnly {
		w += 2
	}
	// 露天／街道：較易談城外與天候（對齊沃土風題材）
	if roomHasTag(mask.RoomTags, "outdoor") {
		if t.ID == "城外聞" {
			w += 5
		}
		if t.ID == "天候" {
			w += 3
		}
	}
	// 人群聚集處：閒聊、八卦更自然
	if roomHasTag(mask.RoomTags, "social") && t.ID == "閒聊" {
		w += 2
	}
	if w < 1 {
		w = 1
	}
	return w
}

// pairRelationWeightAdj 雙人關係、同職場等額外加權。
func pairRelationWeightAdj(t *NpcNpcTopic, familiarity, sentiment int, dyadTags []string) int {
	hasTag := func(s string) bool {
		for _, v := range dyadTags {
			if v == s {
				return true
			}
		}
		return false
	}
	w := 0
	if familiarity >= 50 && t.ID == "閒聊" {
		w += 2
	}
	if hasTag("同職場") && t.ID == "交班" {
		w += 3
	}
	if sentiment <= -20 && t.ID == "打聽" {
		w += 2
	}
	return w
}

// TopicWeightDebug 為單一主題在當前條件下的權重明細（P2.6 debug）。
type TopicWeightDebug struct {
	ID       string `json:"id"`
	Hint     string `json:"hint"`
	Weight   int    `json:"weight"`
	Eligible bool   `json:"eligible"`
	Reason   string `json:"reason,omitempty"`
}

// PickRandomNpcNpcTopicByMask 依情境篩選主題，若無符合則回退為一般隨機（可排除某 id）。
// 權重含 topicMaskBaseWeight（含房間 outdoor 時對「城外聞」「天候」加成）。
func PickRandomNpcNpcTopicByMask(mask NpcTopicMask, excludeID string) *NpcNpcTopic {
	npcTopicsMu.RLock()
	list := npcTopicsList
	npcTopicsMu.RUnlock()
	type weighted struct {
		idx int
		w   int
	}
	var candidates []weighted
	var fallback []int
	for i := range list {
		t := list[i]
		if excludeID != "" && t.ID == excludeID {
			continue
		}
		fallback = append(fallback, i)
		if t.RequiresWork && !mask.IsWorkVenue {
			continue
		}
		if t.NightOnly && !mask.IsNightTime {
			continue
		}
		if t.FollowUp && !mask.HasRoomEvent {
			continue
		}
		w := topicMaskBaseWeight(mask, &t)
		candidates = append(candidates, weighted{idx: i, w: w})
	}
	if len(candidates) > 0 {
		total := 0
		for _, c := range candidates {
			total += c.w
		}
		r := rand.Intn(total)
		acc := 0
		for _, c := range candidates {
			acc += c.w
			if r < acc {
				return &list[c.idx]
			}
		}
	}
	if len(fallback) > 0 {
		return &list[fallback[rand.Intn(len(fallback))]]
	}
	return nil
}

// PickRandomNpcNpcTopicForPair 依情境+關係權重抽主題（P2.5）。
func PickRandomNpcNpcTopicForPair(mask NpcTopicMask, excludeID string, familiarity int, sentiment int, tags []string) *NpcNpcTopic {
	npcTopicsMu.RLock()
	list := npcTopicsList
	npcTopicsMu.RUnlock()
	type weighted struct {
		idx int
		w   int
	}
	var candidates []weighted
	var fallback []int
	for i := range list {
		t := list[i]
		if excludeID != "" && t.ID == excludeID {
			continue
		}
		fallback = append(fallback, i)
		if t.RequiresWork && !mask.IsWorkVenue {
			continue
		}
		if t.NightOnly && !mask.IsNightTime {
			continue
		}
		if t.FollowUp && !mask.HasRoomEvent {
			continue
		}
		w := topicMaskBaseWeight(mask, &t) + pairRelationWeightAdj(&t, familiarity, sentiment, tags)
		if w < 1 {
			w = 1
		}
		candidates = append(candidates, weighted{idx: i, w: w})
	}
	if len(candidates) > 0 {
		total := 0
		for _, c := range candidates {
			total += c.w
		}
		r := rand.Intn(total)
		acc := 0
		for _, c := range candidates {
			acc += c.w
			if r < acc {
				return &list[c.idx]
			}
		}
	}
	if len(fallback) > 0 {
		return &list[fallback[rand.Intn(len(fallback))]]
	}
	return nil
}

// DebugTopicWeightsForPair 回傳主題權重拆解，供 debug API 檢視選題依據（P2.6）。
func DebugTopicWeightsForPair(mask NpcTopicMask, excludeID string, familiarity int, sentiment int, tags []string) []TopicWeightDebug {
	npcTopicsMu.RLock()
	list := npcTopicsList
	npcTopicsMu.RUnlock()
	out := make([]TopicWeightDebug, 0, len(list))
	for _, t := range list {
		d := TopicWeightDebug{ID: t.ID, Hint: t.Hint, Eligible: true, Weight: 0}
		if excludeID != "" && t.ID == excludeID {
			d.Eligible = false
			d.Reason = "excluded by id"
			out = append(out, d)
			continue
		}
		if t.RequiresWork && !mask.IsWorkVenue {
			d.Eligible = false
			d.Reason = "requires_work"
			out = append(out, d)
			continue
		}
		if t.NightOnly && !mask.IsNightTime {
			d.Eligible = false
			d.Reason = "night_only"
			out = append(out, d)
			continue
		}
		if t.FollowUp && !mask.HasRoomEvent {
			d.Eligible = false
			d.Reason = "follow_up without room event"
			out = append(out, d)
			continue
		}
		w := topicMaskBaseWeight(mask, &t) + pairRelationWeightAdj(&t, familiarity, sentiment, tags)
		d.Weight = w
		out = append(out, d)
	}
	return out
}

File: src/npc/topics.rs
// NPC 間對話主題劇本（載入與查詢），對齊 Go npc/topics.go。

use std::fs;
use std::io::ErrorKind;
use std::path::Path;
use std::sync::{LazyLock, RwLock};

use rand::Rng;
use serde::{Deserialize, Serialize};

/// 單一主題劇本（對齊 Go `NpcNpcTopic`）。
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NpcNpcTopic {
    pub id: String,
    pub hint: String,
    #[serde(default)]
    pub lines: Vec<String>,
    #[serde(default)]
    pub requires_work: bool,
    #[serde(default)]
    pub night_only: bool,
    #[serde(default)]
    pub follow_up: bool,
}

#[derive(Debug, Deserialize)]
struct TopicsFile {
    topics: Vec<NpcNpcTopic>,
}

static NPC_TOPICS_LIST: LazyLock<RwLock<Vec<NpcNpcTopic>>> = LazyLock::new(|| RwLock::new(Vec::new()));

/// 從路徑載入主題劇本；檔案不存在則靜默跳過，其餘錯誤寫 log（對齊 Go `LoadNpcNpcTopics`）。
pub fn load_npc_npc_topics(path: impl AsRef<Path>) {
    if let Err(e) = try_load_npc_npc_topics(path.as_ref()) {
        tracing::warn!("[npc_topics] load {}: {}", path.as_ref().display(), e);
    }
}

/// 載入主題；檔案不存在則 `Ok(())` 且不變更列表。
pub fn try_load_npc_npc_topics(path: &Path) -> anyhow::Result<()> {
    let raw = match fs::read_to_string(path) {
        Err(e) if e.kind() == ErrorKind::NotFound => return Ok(()),
        Err(e) => return Err(e.into()),
        Ok(s) => s,
    };
    let f: TopicsFile = serde_json::from_str(&raw)?;
    let list = f.topics;
    let mut g = NPC_TOPICS_LIST.write().expect("npc topics lock");
    *g = list;
    Ok(())
}

/// 依 id 回傳主題（對齊 Go `GetNpcNpcTopicByID`）。
#[must_use]
pub fn get_npc_npc_topic_by_id(id: &str) -> Option<NpcNpcTopic> {
    let g = NPC_TOPICS_LIST.read().expect("npc topics lock");
    g.iter().find(|t| t.id == id).cloned()
}

fn normalize_npc_npc_topic_hint(h: &str) -> String {
    let h = h.trim();
    if let Some(i) = h.find("（路人在旁") {
        return h[..i].trim().to_string();
    }
    h.to_string()
}

/// 依 `topic_hint` 反查主題 id（對齊 Go `FindNpcNpcTopicIDByHint`）。
#[must_use]
pub fn find_npc_npc_topic_id_by_hint(topic_hint: &str) -> String {
    let h = normalize_npc_npc_topic_hint(topic_hint);
    if h.is_empty() {
        return String::new();
    }
    let g = NPC_TOPICS_LIST.read().expect("npc topics lock");
    for t in g.iter() {
        let th = t.hint.trim();
        if !th.is_empty() && h == th {
            return t.id.clone();
        }
    }
    let mut best_id = String::new();
    let mut best_len = 0usize;
    for t in g.iter() {
        let th = t.hint.trim();
        if th.is_empty() {
            continue;
        }
        if h.starts_with(th) && th.len() > best_len {
            best_len = th.len();
            best_id = t.id.clone();
        }
    }
    best_id
}

/// 隨機一主題（對齊 Go `PickRandomNpcNpcTopic`）。
#[must_use]
pub fn pick_random_npc_npc_topic() -> Option<NpcNpcTopic> {
    pick_random_npc_npc_topic_exclude("")
}

/// 隨機主題，可排除 `exclude_id`（對齊 Go `PickRandomNpcNpcTopicExclude`）。
#[must_use]
pub fn pick_random_npc_npc_topic_exclude(exclude_id: &str) -> Option<NpcNpcTopic> {
    let g = NPC_TOPICS_LIST.read().expect("npc topics lock");
    let candidates: Vec<usize> = g
        .iter()
        .enumerate()
        .filter(|(_, t)| exclude_id.is_empty() || t.id != exclude_id)
        .map(|(i, _)| i)
        .collect();
    if candidates.is_empty() {
        return None;
    }
    let mut rng = rand::rng();
    let idx = candidates[rng.random_range(0..candidates.len())];
    g.get(idx).cloned()
}

/// 主題篩選條件（對齊 Go `NpcTopicMask`）。
#[derive(Debug, Clone, Default)]
pub struct NpcTopicMask {
    pub is_work_venue: bool,
    pub is_night_time: bool,
    pub has_room_event: bool,
    pub room_tags: Vec<String>,
}

/// 除錯用權重明細（對齊 Go `TopicWeightDebug`）。
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TopicWeightDebug {
    pub id: String,
    pub hint: String,
    pub weight: i32,
    pub eligible: bool,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub reason: String,
}

#[must_use]
pub fn room_has_tag(room_tags: &[String], want: &str) -> bool {
    room_tags.iter().any(|v| v == want)
}

/// 依情境與房間 tag 計算基礎權重（對齊 Go `topicMaskBaseWeight`）。
#[must_use]
pub fn topic_mask_base_weight(mask: &NpcTopicMask, t: &NpcNpcTopic) -> i32 {
    let mut w = 1;
    if mask.is_work_venue && t.requires_work {
        w += 4;
    }
    if mask.has_room_event && t.follow_up {
        w += 3;
    }
    if mask.is_night_time && t.night_only {
        w += 2;
    }
    if room_has_tag(&mask.room_tags, "outdoor") {
        if t.id == "城外聞" {
            w += 5;
        }
        if t.id == "天候" {
            w += 3;
        }
    }
    if room_has_tag(&mask.room_tags, "social") && t.id == "閒聊" {
        w += 2;
    }
    w.max(1)
}

/// 雙人關係額外加權（對齊 Go `pairRelationWeightAdj`）。
#[must_use]
pub fn pair_relation_weight_adj(t: &NpcNpcTopic, familiarity: i32, sentiment: i32, dyad_tags: &[String]) -> i32 {
    let has_tag = |s: &str| dyad_tags.iter().any(|v| v == s);
    let mut w = 0;
    if familiarity >= 50 && t.id == "閒聊" {
        w += 2;
    }
    if has_tag("同職場") && t.id == "交班" {
        w += 3;
    }
    if sentiment <= -20 && t.id == "打聽" {
        w += 2;
    }
    w
}

/// 依情境加權抽主題；無符合則回退隨機（對齊 Go `PickRandomNpcNpcTopicByMask`）。
#[must_use]
pub fn pick_random_npc_npc_topic_by_mask(mask: &NpcTopicMask, exclude_id: &str) -> Option<NpcNpcTopic> {
    let list: Vec<NpcNpcTopic> = NPC_TOPICS_LIST.read().expect("npc topics lock").clone();
    if list.is_empty() {
        return None;
    }
    let mut candidates: Vec<(usize, i32)> = Vec::new();
    let mut fallback: Vec<usize> = Vec::new();
    for (i, t) in list.iter().enumerate() {
        if !exclude_id.is_empty() && t.id == exclude_id {
            continue;
        }
        fallback.push(i);
        if t.requires_work && !mask.is_work_venue {
            continue;
        }
        if t.night_only && !mask.is_night_time {
            continue;
        }
        if t.follow_up && !mask.has_room_event {
            continue;
        }
        let w = topic_mask_base_weight(mask, t);
        candidates.push((i, w));
    }
    let mut rng = rand::rng();
    if !candidates.is_empty() {
        let total: i32 = candidates.iter().map(|(_, w)| *w).sum();
        if total > 0 {
            let r = rng.random_range(0..total);
            let mut acc = 0;
            for (idx, w) in &candidates {
                acc += w;
                if r < acc {
                    return list.get(*idx).cloned();
                }
            }
        }
    }
    if !fallback.is_empty() {
        let j = rng.random_range(0..fallback.len());
        return list.get(fallback[j]).cloned();
    }
    None
}

/// 情境 + 雙人關係加權抽主題（對齊 Go `PickRandomNpcNpcTopicForPair`）。
#[must_use]
pub fn pick_random_npc_npc_topic_for_pair(
    mask: &NpcTopicMask,
    exclude_id: &str,
    familiarity: i32,
    sentiment: i32,
    tags: &[String],
) -> Option<NpcNpcTopic> {
    let list: Vec<NpcNpcTopic> = NPC_TOPICS_LIST.read().expect("npc topics lock").clone();
    if list.is_empty() {
        return None;
    }
    let mut candidates: Vec<(usize, i32)> = Vec::new();
    let mut fallback: Vec<usize> = Vec::new();
    for (i, t) in list.iter().enumerate() {
        if !exclude_id.is_empty() && t.id == exclude_id {
            continue;
        }
        fallback.push(i);
        if t.requires_work && !mask.is_work_venue {
            continue;
        }
        if t.night_only && !mask.is_night_time {
            continue;
        }
        if t.follow_up && !mask.has_room_event {
            continue;
        }
        let mut w = topic_mask_base_weight(mask, t) + pair_relation_weight_adj(t, familiarity, sentiment, tags);
        if w < 1 {
            w = 1;
        }
        candidates.push((i, w));
    }
    let mut rng = rand::rng();
    if !candidates.is_empty() {
        let total: i32 = candidates.iter().map(|(_, w)| *w).sum();
        if total > 0 {
            let r = rng.random_range(0..total);
            let mut acc = 0;
            for (idx, w) in &candidates {
                acc += w;
                if r < acc {
                    return list.get(*idx).cloned();
                }
            }
        }
    }
    if !fallback.is_empty() {
        let j = rng.random_range(0..fallback.len());
        return list.get(fallback[j]).cloned();
    }
    None
}

/// 除錯：各主題權重拆解（對齊 Go `DebugTopicWeightsForPair`）。
#[must_use]
pub fn debug_topic_weights_for_pair(
    mask: &NpcTopicMask,
    exclude_id: &str,
    familiarity: i32,
    sentiment: i32,
    tags: &[String],
) -> Vec<TopicWeightDebug> {
    let g = NPC_TOPICS_LIST.read().expect("npc topics lock");
    let mut out = Vec::with_capacity(g.len());
    for t in g.iter() {
        let mut d = TopicWeightDebug {
            id: t.id.clone(),
            hint: t.hint.clone(),
            eligible: true,
            weight: 0,
            reason: String::new(),
        };
        if !exclude_id.is_empty() && t.id == exclude_id {
            d.eligible = false;
            d.reason = "excluded by id".to_string();
            out.push(d);
            continue;
        }
        if t.requires_work && !mask.is_work_venue {
            d.eligible = false;
            d.reason = "requires_work".to_string();
            out.push(d);
            continue;
        }
        if t.night_only && !mask.is_night_time {
            d.eligible = false;
            d.reason = "night_only".to_string();
            out.push(d);
            continue;
        }
        if t.follow_up && !mask.has_room_event {
            d.eligible = false;
            d.reason = "follow_up without room event".to_string();
            out.push(d);
            continue;
        }
        d.weight = topic_mask_base_weight(mask, t) + pair_relation_weight_adj(t, familiarity, sentiment, tags);
        out.push(d);
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn topics_path() -> std::path::PathBuf {
        std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("data/npc_to_npc_topics.json")
    }

    #[test]
    fn find_topic_id_by_hint_matches_go() {
        try_load_npc_npc_topics(&topics_path()).expect("load topics");
        assert_eq!(
            find_npc_npc_topic_id_by_hint("換班、交接時的情境，簡短交代或叮囑"),
            "交班"
        );
        assert_eq!(
            find_npc_npc_topic_id_by_hint("隨口閒聊，天氣、見聞、抱怨（路人在旁，兩句務必極短。）"),
            "閒聊"
        );
        assert_eq!(
            find_npc_npc_topic_id_by_hint(
                "鎮上傳聞城外惡地、霜林、游離輻射或富態化巨獸（非荒蕪末世，而是生機過剩的危險）"
            ),
            "城外聞"
        );
        assert_eq!(find_npc_npc_topic_id_by_hint(""), "");
    }

    #[test]
    fn get_by_id_after_load() {
        try_load_npc_npc_topics(&topics_path()).expect("load");
        assert_eq!(
            get_npc_npc_topic_by_id("閒聊").map(|t| t.hint.contains("隨口")),
            Some(true)
        );
    }
}
